tests/unit/drivers/test_recursive_traversal.py

```diff
@@ -198,3 +198,43 @@ def test_adjacency1_granularity1():
     assert len(docs[0].matches[0].matches) == DOCUMENTS_PER_LEVEL
     assert len(docs[0].matches[0].matches[0].chunks) == DOCUMENTS_PER_LEVEL
 
+
+def test_adjacency_on_chunks():
+    docs = build_docs()
+    doc = docs[0]
+    doc.ClearField('matches')
+    docs = [doc, ]
+    driver = SliceQL(
+        start=0,
+        end=1,
+        adjacency_range=(1, 1),
+        granularity_range=(1, 2),
+        recur_on=["chunks", ]
+    )
+    driver._traverse_apply(docs)
+    assert len(docs) == 1
+    assert len(docs[0].chunks) == 1
+    assert len(docs[0].chunks[0].chunks) == 1
+    assert len(docs[0].chunks[0].chunks[0].matches) == DOCUMENTS_PER_LEVEL
+    assert len(docs[0].chunks[0].matches) == DOCUMENTS_PER_LEVEL
+
+
+def test_granularity_on_matches():
+    docs = build_docs()
+    doc = docs[0]
+    doc.ClearField('chunks')
+    docs = [doc, ]
+    driver = SliceQL(
+        start=0,
+        end=1,
+        adjacency_range=(1, 2),
+        granularity_range=(0, 1),
+        recur_on=["matches", ]
+    )
+    driver._traverse_apply(docs)
+    assert len(docs) == 1
+    assert len(docs[0].matches) == 1
```

Comment:
In contrast to the upper example: This is correct, since you set `granularity_range=(0,1)`, e.g. touch the first adjacency layer on the lowest chunk layer. Anyhow, the `break` statement in the `_traversal` function is not called, since the `_granularity_start` is `0` and thus there is no need for zooming into the structure.

Reply:
Yes, you are right. But still, the current implementation is limited. I've updated the test to illustrate my points.